README.md

```diff
@@ -5,26 +5,77 @@
 [![License](https://img.shields.io/cocoapods/l/swift-async-queue.svg)](https://cocoapods.org/pods/swift-async-queue)
 [![Platform](https://img.shields.io/cocoapods/p/swift-async-queue.svg)](https://cocoapods.org/pods/swift-async-queue)
 
-A queue that enables sending FIFO-ordered tasks from synchronous to asynchronous contexts.
+A library of queues that enables sending ordered tasks from synchronous to asynchronous contexts.
 
-## Usage
+## Task Ordering and Swift Concurrency
 
-### Basic Initialization
+Tasks sent from a synchronous context to an asynchronous context in Swift Concurrency are inherently unordered. Consider the following test:
 
-```swift
-let asyncQueue = AsyncQueue()
 ```
+@MainActor
+func test_mainActor_taskOrdering() async {
+    var counter = 0
+    var tasks = [Task<Void, Never>]()
+    for iteration in 1...100 {
+        tasks.append(Task {
+            counter += 1
+            XCTAssertEqual(counter, iteration) // often fails
+        })
+    }
+    for task in tasks {
+        _ = await task.value
+    }
+}
+```
+
+Despite the spawned `Task` inheriting the serial `@MainActor` execution context, the ordering of the scheduled asynchronous work is not guaranteed.
 
-### Sending events from a synchronous context
+While [actors](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html#ID645) are great at serializing tasks, there is no simple way in the standard Swift library to send ordered tasks to them from a synchronous context.
+
+### Executing asynchronous tasks in FIFO order
+
+Use a `FIFOQueue` queue to execute asynchronous tasks enqueued from a nonisolated context in FIFO order. Tasks sent to one of these queues are guaranteed to begin _and end_ executing in the order in which they are enqueued.
 
 ```swift
-asyncQueue.async { /* awaitable context that executes after all other enqueued work is completed */ }
+let queue = FIFOQueue()
+queue.async {
+    /*
+    `async` context that executes after all other enqueued work is completed.
+    Work enqueued after this task will wait for this task to complete.
+    */
+}
+Task {
+    await queue.await {
+        /*
+        `async` context that can return a value or throw an error.
+        Executes after all other enqueued work is completed.
+        Work enqueued after this task will wait for this task to complete.
+        */
+    }
+}
 ```
 
-### Awaiting work from an asynchronous context
+### Sending ordered asynchronous tasks to Actors
+
+Use an `ActorQueue` queue to send ordered asynchronous tasks from a nonisolated context to an `actor` instance. Tasks sent to one of these queues are guaranteed to begin _but not end_ executing in the order in which they are enqueued.
 
 ```swift
-await asyncQueue.await { /* throw-able, return-able, awaitable context that executes after all other enqueued work is completed */ }
+let queue = ActorQueue()
+queue.async {
+    /*
+    `async` context that executes after all other enqueued work has begun executing.
```

Comment:
Is it possible to call non-async actor methods synchronously from the outside, by setting us in the same execution context or something like that?

Actors have access to their own serial executors so it may work out to have some kind of actor extension that runs non-serial code through the executor. Although Swift doesn't let us extend a protocol so we'd need to put some ornamentation around such an idea.

I'll need to discuss this further, I'm sure you've been giving this a lot of thought as well.

Reply:
> Is it possible to call non-async actor methods synchronously from the outside, by setting us in the same execution context or something like that?

Yes, if you use [`@_inheritActorContext`](https://github.com/apple/swift/blob/1f3e159cfe518fbb723230a4e8f562248d34d863/docs/ReferenceGuides/UnderscoredAttributes.md#_inheritactorcontext). But given the `Usage of these attributes outside of the Swift monorepo is STRONGLY DISCOURAGED` warning at the top of that link, I do not plan on using underscored attributes in this repo.

> Actors have access to their own serial executors so it may work out to have some kind of actor extension that runs non-serial code through the executor.

This would be ideal. I haven't found a supported way to do this though.

> Although Swift doesn't let us extend a protocol so we'd need to put some ornamentation around such an idea.

Indeed we'd need to at least be able to store the `taskStreamContinuation`, and we can't do that on a protocol extension. Worse yet, `actor`s do not support inheritance, so we're quite low on options here.

Reply:
Following up: I don't think our use case for an `ActorQueue` would result in us wanting to inherit the actor context. The entire point of an `ActorQueue` is to bridge the nonisolated context with the actor-isolated context – inheriting an actor context from a nonisolated context doesn't improve the semantics since we're calling into the `ActorQueue` from a context other than the actor's.